kvclient: replace ReplicaSlice with ReplicaSet for routing

andrewbaptist opened the pull request:
This PR replaces ReplicaSlice with ReplicaSet outside of the DistSender. ReplicaSlice is an internal implementation which contains information only required for sorting the replicas. Outside of DistSender the additional sorting information is unnecessary and unused.

Epic: none
Informs: #112351

Release note: None

Commit:
kvclient: replace ReplicaSlice with ReplicaSet for routing

This PR replaces ReplicaSlice with ReplicaSet outside of the
DistSender. ReplicaSlice is an internal implementation which
contains information only required for sorting the replicas.
Outside of DistSender the additional sorting information is
unnecessary and unused.

Epic: none
Informs: #112351

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114429)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2307 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114429#-NjJuXFP0grEfPPmty20:-NjJuXFQC4FaVfiESLuG:b15tsji) ([raw file](https://github.com/cockroachdb/cockroach/blob/44449e57088fb97844b76461625082b3e8427423/pkg/kv/kvclient/kvcoord/dist_sender.go#L2307)):*
> ```Go
> 		dontConsiderConnHealth: ds.dontConsiderConnHealth,
> 	}
> 	transport, err := ds.transportFactory(opts, ds.nodeDialer, replicas.AsReplicaSet())
> ```

Does this new heap allocation (in `ReplicaSlice.Descriptors`) on the request path show up in microbenchmarks?


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2307 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114429#-NjJuXFP0grEfPPmty20:-NjNvm249CbFtkIX8EyE:b5n4lh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/44449e57088fb97844b76461625082b3e8427423/pkg/kv/kvclient/kvcoord/dist_sender.go#L2307)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this new heap allocation (in `ReplicaSlice.Descriptors`) on the request path show up in microbenchmarks?
</blockquote></details>

I didn't explicitly test this. The "following" PR (#114033) removes this copy. I wasn't sure if it was easiest to do this in 1 or 2 PRs, but I understand the concern if this was the last PR. Let me know if you think its worth testing, or if I should close this PR and replace with the other one only.


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
No noticible bench perf change (likely just noise) and no additional allocs. This change makes it easier to cache these going forward which will greatly improve the performance of this test as about 25-35% of CPU and 90% of memory deals with allocating, sorting and managing the ReplicaSlice/Set.

Running this command before and after this commit:
```
./dev bench pkg/kv/kvclient/kvcoord -f BenchmarkDistSenderSunnyDay -v --ignore-cache --bench-mem --bench-time 1s --timeout 500s
```

Before:
```
BenchmarkDistSenderSunnyDay/rf-3/count-1-conc-10         	  943327	      1276 ns/op	    1688 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-1-10              	  607777	      1793 ns/op	    1688 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-1000-conc-10      	  839977	      1345 ns/op	    1687 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-1000-10           	  551778	      2025 ns/op	    1688 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-100000-conc-10    	  950823	      1411 ns/op	    1687 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-100000-10         	  461899	      2348 ns/op	    1688 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-1-conc-10         	  957303	      1328 ns/op	    1847 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-1-10              	  596076	      1933 ns/op	    1848 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-1000-conc-10      	  831895	      1358 ns/op	    1847 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-1000-10           	  501507	      2223 ns/op	    1848 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-100000-conc-10    	  825180	      1453 ns/op	    1847 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-100000-10         	  488252	      2515 ns/op	    1848 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-1-conc-10        	  820792	      1447 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-1-10             	  445353	      2532 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-1000-conc-10     	  814402	      1469 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-1000-10          	  412800	      2836 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-100000-conc-10   	  789969	      1508 ns/op	    2327 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-100000-10        	  329214	      3094 ns/op	    2328 B/op	      19 allocs/op
```
After:
```
BenchmarkDistSenderSunnyDay/rf-3/count-1-conc-10         	  934548	      1280 ns/op	    1687 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-1-10              	  610624	      1815 ns/op	    1688 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-1000-conc-10      	  864946	      1309 ns/op	    1688 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-1000-10           	  545544	      2052 ns/op	    1688 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-100000-conc-10    	  841906	      1414 ns/op	    1687 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-3/count-100000-10         	  515714	      2335 ns/op	    1688 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-1-conc-10         	  963768	      1321 ns/op	    1848 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-1-10              	  572091	      1958 ns/op	    1848 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-1000-conc-10      	  834961	      1345 ns/op	    1847 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-1000-10           	  505413	      2227 ns/op	    1848 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-100000-conc-10    	  844149	      1427 ns/op	    1847 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-5/count-100000-10         	  491342	      2490 ns/op	    1848 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-1-conc-10        	  813049	      1410 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-1-10             	  433617	      2541 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-1000-conc-10     	  820797	      1424 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-1000-10          	  412074	      2822 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-100000-conc-10   	  917679	      1582 ns/op	    2328 B/op	      19 allocs/op
BenchmarkDistSenderSunnyDay/rf-11/count-100000-10        	  359485	      3192 ns/op	    2328 B/op	      19 allocs/op
```

andrewbaptist commented:
@nvanbenschoten can you take another look at this now that I added the microbenchmark. The next step will be to cache these and by making this change, the future change is smaller. 

I added @arulajmani as a reviewer as well.

nvb reviewed (approved):
> No noticible bench perf change (likely just noise) and no additional allocs.

I don't really see how there isn't any change in allocations, given the new call to `ReplicaSlice.AsReplicaSet`. Is the benchmark measuring the performance of `sendToReplicas`?

Regardless, if the plan is to cache these going forward, then this change <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, as long as we still intend to land that change in v24.1.

> Reviewed 1 of 15 files at r1, 15 of 15 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114429#-:-Ns8wgWu6pNe9KJrffUA:bmdlzst)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist, @arulajmani, @msbutler, and @yuzefovich)

___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114429#-Ns8vS5X9NZP3HZiKHZV:-Ns8vS5Y3ANsPuw7DP2H:bsomucw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9ebaafc65507579f5fbded0454e0a6b4482be6/pkg/kv/kvclient/kvcoord/replica_slice.go#L269)):*
> ```Go
> }
> 
> func (rs ReplicaSlice) AsReplicaSet() roachpb.ReplicaSet {
> ```

This method is exported, so it needs a comment. There should be a linter that's firing.
___
*[`pkg/roachpb/metadata_replicas.go` line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114429#-Ns8vg8rDxTfgrG2ozCH:-Ns8vg8rDxTfgrG2ozCI:b-xhs14s) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9ebaafc65507579f5fbded0454e0a6b4482be6/pkg/roachpb/metadata_replicas.go#L50)):*
> ```Go
> }
> 
> func (d ReplicaSet) First() ReplicaDescriptor {
> ```

Same point about needing a comment on the exported method.


<!-- Sent from Reviewable.io -->
